File: libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/files/FileRequests.java
/*
 * Copyright (c) 2013-2014, salesforce.com, inc.
 * All rights reserved.
 * Redistribution and use of this software in source and binary forms, with or
 * without modification, are permitted provided that the following conditions
 * are met:
 * - Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * - Neither the name of salesforce.com, inc. nor the names of its contributors
 * may be used to endorse or promote products derived from this software without
 * specific prior written permission of salesforce.com, inc.
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.salesforce.androidsdk.rest.files;

import android.text.TextUtils;
import android.util.Log;

import com.google.common.collect.Maps;
import com.salesforce.androidsdk.app.SalesforceSDKManager;
import com.salesforce.androidsdk.auth.HttpAccess;
import com.salesforce.androidsdk.rest.ApiVersionStrings;
import com.salesforce.androidsdk.rest.RestClient;
import com.salesforce.androidsdk.rest.RestRequest;
import com.salesforce.androidsdk.rest.RestRequest.RestMethod;
import com.salesforce.androidsdk.rest.RestResponse;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.HttpVersion;
import org.apache.http.ProtocolVersion;
import org.apache.http.StatusLine;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.BasicHttpEntity;
import org.apache.http.entity.StringEntity;
import org.apache.http.message.BasicHttpResponse;
import org.apache.http.message.BasicStatusLine;
import org.apache.http.protocol.HTTP;

import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.util.List;
import java.util.Map;

/**
 * This defines the HTTP requests in the connect API for files functionality.
 * 
 * @author sfell
 */
public class FileRequests extends ApiRequests {

    // files i follow
    static final String CONTENT_DOCUMENT_LINK = ApiVersionStrings.BASE_SOBJECT_PATH + "ContentDocumentLink";

    /**
     * Build a Request that can fetch a page from the files owned by the
     * specified user.
     * 
     * @param userId
     *            if null the context user is used, otherwise it should be an Id
     *            of a user.
     * @param pageNum
     *            if null fetches the first page, otherwise fetches the
     *            specified page.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest ownedFilesList(String userId, Integer pageNum) {
        return make(base("users").appendUserId(userId).appendPath("files").appendPageNum(pageNum));
    }

    /**
     * Build a Request that can fetch a page from the list of files from groups
     * that the user is a member of.
     * 
     * @param userId
     *            if null the context user is used, otherwise it should be an Id
     *            of a user.
     * @param pageNum
     *            if null fetches the first page, otherwise fetches the
     *            specified page.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest filesInUsersGroups(String userId, Integer pageNum) {
        return make(base("users").appendUserId(userId).appendPath("files/filter/groups").appendPageNum(pageNum));
    }

    /**
     * Build a Request that can fetch a page from the list of files that have
     * been shared with the user.
     * 
     * @param userId
     *            if null the context user is used, otherwise it should be an Id
     *            of a user.
     * @param pageNum
     *            if null fetches the first page, otherwise fetches the
     *            specified page.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest filesSharedWithUser(String userId, Integer pageNum) {
        return make(base("users").appendUserId(userId).appendPath("files/filter/sharedwithme").appendPageNum(pageNum));
    }

    /**
     * Build a Request that can fetch the file details of a particular version
     * of a file.
     * 
     * @param sfdcId
     *            The Id of the file
     * @param version
     *            if null fetches the most recent version, otherwise fetches
     *            this specific version.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest fileDetails(String sfdcId, String version) {
        validateSfdcId(sfdcId);
        return make(base("files").appendPath(sfdcId).appendVersionNum(version));
    }

    /**
     * Build a request that can fetch the latest file details of one or more
     * files in a single request.
     * 
     * @param sfdcIds
     *            The list of file Ids to fetch.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest batchFileDetails(List<String> sfdcIds) {
        validateSfdcIds(sfdcIds);
        String ids = TextUtils.join(",", sfdcIds);
        return make(base("files").appendPath("batch").appendPath(ids));
    }

    /**
     * Build a Request that can fetch the a preview/rendition of a particular
     * page of the file (and version)
     * 
     * @param sfdcId
     *            The Id of the file
     * @param version
     *            if null fetches the most recent version, otherwise fetches
     *            this specific version
     * @param renditionType
     *            What format of rendition do you want to get
     * @param pageNum
     *            which page to fetch, pages start at 0.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest fileRendition(String sfdcId, String version, RenditionType renditionType, Integer pageNum) {
        validateSfdcId(sfdcId);
        if (renditionType == null) {
            throw new NullPointerException("rendition type can't be null");
        }
        return make(base("files").appendPath(sfdcId).appendPath("rendition")
                .appendQueryParam("type", renditionType.toString()).appendVersionNum(version).appendPageNum(pageNum));
    }

    /**
     * Builds a request that can fetch the actual binary file contents of this
     * particular file.
     * 
     * @param sfdcId
     *            The Id of the file
     * @param version
     *            The version of the file
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest fileContents(String sfdcId, String version) {
        validateSfdcId(sfdcId);
        return make(base("files").appendPath(sfdcId).appendPath("content").appendVersionNum(version));
    }

    /**
     * Build a Request that can fetch a page from the list of entities that this
     * file is shared to.
     * 
     * @param sfdcId
     *            The Id of the file.
     * @param pageNum
     *            if null fetches the first page, otherwise fetches the
     *            specified page.
     * @return A new RestRequest that can be used to fetch this data
     */
    public static RestRequest fileShares(String sfdcId, Integer pageNum) {
        validateSfdcId(sfdcId);
        return make(base("files").appendPath(sfdcId).appendPath("file-shares").appendPageNum(pageNum));
    }

    /**
     * Build a request that will add a file share for the specified fileId to
     * the specified entityId
     * 
     * @param fileId
     *            the Id of the file being shared.
     * @param entityId
     *            the Id of the entity to share the file to (e.g. a user or a
     *            group)
     * @param shareType
     *            the type of share (V - View, C - Collaboration)
     * 
     * @return A new RestRequest that be used to create this share.
     */
    public static RestRequest addFileShare(String fileId, String entityId, String shareType) {
        validateSfdcIds(fileId, entityId);
        return new RestRequest(RestMethod.POST, CONTENT_DOCUMENT_LINK, makeFileShare(fileId, entityId, shareType));
    }

    /**
     * Build a request that will delete the specified file share.
     * 
     * @param shareId
     *            The Id of the file share record (aka ContentDocumentLink)
     * 
     * @return
     */
    public static RestRequest deleteFileShare(String shareId) {
        validateSfdcId(shareId);
        return new RestRequest(RestMethod.DELETE, CONTENT_DOCUMENT_LINK + "/" + shareId, null);
    }

    private static HttpEntity makeFileShare(String fileId, String entityId, String shareType) {
        Map<String, String> share = Maps.newHashMap();
        share.put("ContentDocumentId", fileId);
        share.put("LinkedEntityId", entityId);
        share.put("ShareType", shareType);
        try {
            String json = null; // FIXME ConnectParser.toJson(share);
            StringEntity e = new StringEntity(json, HTTP.UTF_8);
            e.setContentType("application/json");
            return e;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: flutter_sdk/lib/src/core/rest/file_requests.dart
import 'rest_request.dart';

/// Types of file renditions supported by the server.
enum RenditionType {
  flash,
  pdf,
  thumb120by90,
  thumb240by180,
  thumb720by480;

  /// Whether this rendition type produces a PNG image.
  bool get isPng =>
      this == thumb120by90 || this == thumb240by180 || this == thumb720by480;

  @override
  String toString() => name.toUpperCase();
}

/// Provides factory methods for Salesforce Chatter Files REST API requests.
///
/// Supports listing, viewing, sharing, and uploading files through
/// the Connect API.
class FileRequests {
  static const String _servicesData = RestRequest.servicesData;
  static const String _defaultApiVersion = '62.0';

  /// Returns the base path for ContentDocumentLink operations.
  static String getContentDocumentLinkPath({String apiVersion = _defaultApiVersion}) {
    return '$_servicesData/v$apiVersion/sobjects/ContentDocumentLink';
  }

  /// Builds a request to fetch files owned by the specified user.
  ///
  /// [userId] If null, the context user is used.
  /// [pageNum] If null, fetches the first page.
  static RestRequest ownedFilesList({
    String apiVersion = _defaultApiVersion,
    String? userId,
    int? pageNum,
  }) {
    var path = '$_servicesData/v$apiVersion/connect/files/users';
    path = _appendUserId(path, userId);
    path = _appendPageNum(path, pageNum);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to fetch files from groups that the user is a member of.
  ///
  /// [userId] If null, the context user is used.
  /// [pageNum] If null, fetches the first page.
  static RestRequest filesInUsersGroups({
    String apiVersion = _defaultApiVersion,
    String? userId,
    int? pageNum,
  }) {
    var path = '$_servicesData/v$apiVersion/connect/files/users';
    path = _appendUserId(path, userId);
    path += '/filter/groups';
    path = _appendPageNum(path, pageNum);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to fetch files shared with the specified user.
  ///
  /// [userId] If null, the context user is used.
  /// [pageNum] If null, fetches the first page.
  static RestRequest filesSharedWithUser({
    String apiVersion = _defaultApiVersion,
    String? userId,
    int? pageNum,
  }) {
    var path = '$_servicesData/v$apiVersion/connect/files/users';
    path = _appendUserId(path, userId);
    path += '/filter/sharedwithme';
    path = _appendPageNum(path, pageNum);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to fetch file details for a specific version.
  ///
  /// [sfdcId] The ID of the file.
  /// [version] If null, fetches the most recent version.
  static RestRequest fileDetails({
    required String apiVersion,
    required String sfdcId,
    String? version,
  }) {
    _validateSfdcId(sfdcId);
    var path = '$_servicesData/v$apiVersion/connect/files/$sfdcId';
    path = _appendVersionNum(path, version);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to fetch details of multiple files in a single call.
  ///
  /// [sfdcIds] The list of file IDs to fetch.
  static RestRequest batchFileDetails({
    String apiVersion = _defaultApiVersion,
    required List<String> sfdcIds,
  }) {
    _validateSfdcIds(sfdcIds);
    final ids = sfdcIds.join(',');
    return RestRequest(
      method: RestMethod.get,
      path: '$_servicesData/v$apiVersion/connect/files/batch/$ids',
    );
  }

  /// Builds a request to fetch a preview/rendition of a file page.
  ///
  /// [sfdcId] The ID of the file.
  /// [version] If null, fetches the most recent version.
  /// [renditionType] The format of rendition to get.
  /// [pageNum] Which page to fetch (pages start at 0).
  static RestRequest fileRendition({
    String apiVersion = _defaultApiVersion,
    required String sfdcId,
    String? version,
    required RenditionType renditionType,
    int? pageNum,
  }) {
    _validateSfdcId(sfdcId);
    var path =
        '$_servicesData/v$apiVersion/connect/files/$sfdcId/rendition?type=$renditionType';
    path = _appendVersionNum(path, version);
    path = _appendPageNum(path, pageNum);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to fetch the binary contents of a file.
  ///
  /// [sfdcId] The ID of the file.
  /// [version] The version of the file.
  static RestRequest fileContents({
    String apiVersion = _defaultApiVersion,
    required String sfdcId,
    String? version,
  }) {
    _validateSfdcId(sfdcId);
    var path = '$_servicesData/v$apiVersion/connect/files/$sfdcId/content';
    path = _appendVersionNum(path, version);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to fetch the entities a file is shared with.
  ///
  /// [sfdcId] The ID of the file.
  /// [pageNum] If null, fetches the first page.
  static RestRequest fileShares({
    String apiVersion = _defaultApiVersion,
    required String sfdcId,
    int? pageNum,
  }) {
    _validateSfdcId(sfdcId);
    var path =
        '$_servicesData/v$apiVersion/connect/files/$sfdcId/file-shares';
    path = _appendPageNum(path, pageNum);
    return RestRequest(method: RestMethod.get, path: path);
  }

  /// Builds a request to share a file with an entity.
  ///
  /// [fileId] The ID of the file being shared.
  /// [entityId] The ID of the entity to share with (user or group).
  /// [shareType] The type of share ('V' for View, 'C' for Collaboration).
  static RestRequest addFileShare({
    String apiVersion = _defaultApiVersion,
    required String fileId,
    required String entityId,
    required String shareType,
  }) {
    _validateSfdcId(fileId);
    _validateSfdcId(entityId);
    return RestRequest(
      method: RestMethod.post,
      path: getContentDocumentLinkPath(apiVersion: apiVersion),
      requestBody: {
        'ContentDocumentId': fileId,
        'LinkedEntityId': entityId,
        'ShareType': shareType,
      },
    );
  }

  /// Builds a request to delete a file share.
  ///
  /// [shareId] The ID of the file share record (ContentDocumentLink).
  static RestRequest deleteFileShare({
    String apiVersion = _defaultApiVersion,
    required String shareId,
  }) {
    _validateSfdcId(shareId);
    return RestRequest(
      method: RestMethod.delete,
      path: '${getContentDocumentLinkPath(apiVersion: apiVersion)}/$shareId',
    );
  }

  /// Builds a multipart request to upload a file.
  ///
  /// Note: In Flutter, file upload requires using a multipart HTTP request.
  /// This method builds the request metadata; the caller is responsible for
  /// attaching the file data as a multipart body using their HTTP client.
  ///
  /// [name] The file name.
  /// [title] The title of the file.
  /// [description] A description of the file.
  static RestRequest uploadFile({
    String apiVersion = _defaultApiVersion,
    required String name,
    String? title,
    String? description,
  }) {
    final body = <String, dynamic>{};
    if (title != null) body['title'] = title;
    if (description != null) body['desc'] = description;
    body['fileName'] = name;
    return RestRequest(
      method: RestMethod.post,
      path: '$_servicesData/v$apiVersion/connect/files/users/me',
      requestBody: body,
    );
  }

  static String _appendUserId(String path, String? userId) {
    return '$path/${userId ?? 'me'}';
  }

  static String _appendPageNum(String path, int? pageNum) {
    if (pageNum == null) return path;
    final separator = path.contains('?') ? '&' : '?';
    return '$path${separator}page=$pageNum';
  }

  static String _appendVersionNum(String path, String? version) {
    if (version == null) return path;
    final separator = path.contains('?') ? '&' : '?';
    return '$path${separator}versionNumber=$version';
  }

  static void _validateSfdcId(String id) {
    if (id.isEmpty) {
      throw ArgumentError('Salesforce ID cannot be empty');
    }
  }

  static void _validateSfdcIds(List<String> ids) {
    if (ids.isEmpty) {
      throw ArgumentError('Salesforce ID list cannot be empty');
    }
    for (final id in ids) {
      _validateSfdcId(id);
    }
  }
}
